File: Frotas-client-api/Frotas.API.Application/Utility/TipoViaturaHelper.cs
using System;
using Frotas.API.Domain.Entities.Frotas;

namespace Frotas.API.Application.Utility
{
  /// <summary>
  /// Helper para mapear designações de tipos de viatura para categorias de inspeção
  /// Porquê: Permite calcular automaticamente a categoria de inspeção baseado na designação escolhida
  /// </summary>
  public static class TipoViaturaHelper
  {
    /// <summary>
    /// Mapeia a designação do tipo de viatura para a categoria de inspeção
    /// </summary>
    /// <param name="designacao">Designação do tipo de viatura</param>
    /// <returns>Categoria de inspeção correspondente ou Ligeiro como padrão</returns>
    public static CategoriaInspecao MapearDesignacaoParaCategoriaInspecao(string? designacao)
    {
      if (string.IsNullOrWhiteSpace(designacao))
      {
        return CategoriaInspecao.Ligeiro; // Valor padrão
      }

      string designacaoLower = designacao.ToLower().Trim();

      // Mapear designações para categorias de inspeção
      if (designacaoLower == "ligeiro passageiros")
      {
        return CategoriaInspecao.Ligeiro;
      }

      if (designacaoLower == "ligeiro mercadorias")
      {
        return CategoriaInspecao.LigeiroMercadorias;
      }

      if (designacaoLower == "pesado")
      {
        return CategoriaInspecao.Pesado;
      }

      // Motociclo e tratores agrícolas não têm regras específicas de inspeção definidas
      // Retornar Ligeiro como padrão (será tratado como anual no helper de inspeção)
      if (designacaoLower == "motociclo" || designacaoLower == "tratores agricolas")
      {
        return CategoriaInspecao.Ligeiro; // Usar Ligeiro como padrão, mas será tratado como anual
      }

      // Valor padrão para designações desconhecidas
      return CategoriaInspecao.Ligeiro;
    }
  }
}

File: Frotas-client-frontend/src/utils/tipo-viatura-helper.ts
import type { CategoriaInspecao } from '@/types/dtos/frotas/tipo-viaturas.dtos'

/**
 * Designações pré-definidas para tipos de viatura
 * Porquê: Valores fixos que o utilizador pode escolher ao criar/editar um tipo de viatura
 */
export const TIPO_VIATURA_DESIGNACOES = [
  'ligeiro passageiros',
  'ligeiro mercadorias',
  'pesado',
  'motociclo',
  'tratores agricolas',
] as const

export type TipoViaturaDesignacao = (typeof TIPO_VIATURA_DESIGNACOES)[number]

/**
 * Mapeia a designação do tipo de viatura para a categoria de inspeção
 * Porquê: Permite calcular as datas de inspeção baseado na designação escolhida
 * 
 * @param designacao Designação do tipo de viatura
 * @returns Categoria de inspeção correspondente ou undefined se não mapeado
 */
export function mapearDesignacaoParaCategoriaInspecao(
  designacao: string | null | undefined
): CategoriaInspecao | undefined {
  if (!designacao) {
    return undefined
  }

  const designacaoLower = designacao.toLowerCase().trim()

  // Mapear designações para categorias de inspeção
  // Verificar variações possíveis da designação
  if (
    designacaoLower === 'ligeiro passageiros' ||
    designacaoLower.includes('ligeiro') && designacaoLower.includes('passageiro')
  ) {
    return 'ligeiro'
  }

  if (
    designacaoLower === 'ligeiro mercadorias' ||
    designacaoLower.includes('ligeiro') && designacaoLower.includes('mercadoria')
  ) {
    return 'ligeiroMercadorias'
  }

  if (designacaoLower === 'pesado' || designacaoLower.includes('pesado')) {
    return 'pesado'
  }

  // Motociclo e tratores agrícolas não têm regras específicas de inspeção definidas
  // Retornar undefined para usar regra padrão (anual)
  if (
    designacaoLower === 'motociclo' ||
    designacaoLower.includes('motociclo') ||
    designacaoLower === 'tratores agricolas' ||
    designacaoLower.includes('trator') ||
    designacaoLower.includes('agricola')
  ) {
    return undefined
  }

  // Se não encontrou correspondência exata, tentar inferir baseado em palavras-chave
  if (designacaoLower.includes('pesado')) {
    return 'pesado'
  }
  if (designacaoLower.includes('mercadoria')) {
    return 'ligeiroMercadorias'
  }
  if (designacaoLower.includes('ligeiro')) {
    return 'ligeiro'
  }

  return undefined
}

